File: scala/src/main/scala/algorithms/QuickSort.scala
package algorithms

import scala.reflect.ClassTag

/**
 * Quick sort with median-of-three pivot selection.
 * Time: O(n log n) average, O(n²) worst case
 * Space: O(log n) stack space
 */
object QuickSort:
  def sort[T: ClassTag](arr: Array[T])(using ord: Ordering[T]): Array[T] =
    if arr.length <= 1 then arr
    else
      quickSort(arr, 0, arr.length - 1)
      arr

  private def quickSort[T](arr: Array[T], low: Int, high: Int)(using ord: Ordering[T]): Unit =
    if low < high then
      val pivotIndex = partition(arr, low, high)
      quickSort(arr, low, pivotIndex - 1)
      quickSort(arr, pivotIndex + 1, high)

  private def partition[T](arr: Array[T], low: Int, high: Int)(using ord: Ordering[T]): Int =
    val pivotIndex = medianOfThree(arr, low, high)
    swap(arr, pivotIndex, high)
    val pivot = arr(high)

    var i = low - 1
    for j <- low until high do
      if ord.lteq(arr(j), pivot) then
        i += 1
        swap(arr, i, j)

    swap(arr, i + 1, high)
    i + 1

  private def medianOfThree[T](arr: Array[T], low: Int, high: Int)(using ord: Ordering[T]): Int =
    val mid = low + (high - low) / 2

    if ord.gt(arr(low), arr(mid)) then swap(arr, low, mid)
    if ord.gt(arr(low), arr(high)) then swap(arr, low, high)
    if ord.gt(arr(mid), arr(high)) then swap(arr, mid, high)

    mid

  private def swap[T](arr: Array[T], i: Int, j: Int): Unit =
    val temp = arr(i)
    arr(i) = arr(j)
    arr(j) = temp

  def sorted[T: ClassTag](arr: Array[T])(using ord: Ordering[T]): Array[T] =
    val copy = arr.clone()
    sort(copy)

File: zig/src/algorithms/quick_sort.zig
const std = @import("std");

/// Quick sort with median-of-three pivot selection.
/// Time: O(n log n) average, O(n²) worst case
/// Space: O(log n) stack space

pub fn sort(comptime T: type, arr: []T) void {
    if (arr.len <= 1) return;
    quickSort(T, arr, 0, arr.len - 1);
}

fn quickSort(comptime T: type, arr: []T, low: usize, high: usize) void {
    if (low >= high) return;

    const pivot_index = partition(T, arr, low, high);

    if (pivot_index > 0) {
        quickSort(T, arr, low, pivot_index - 1);
    }
    quickSort(T, arr, pivot_index + 1, high);
}

fn partition(comptime T: type, arr: []T, low: usize, high: usize) usize {
    const pivot_index = medianOfThree(T, arr, low, high);
    const temp = arr[pivot_index];
    arr[pivot_index] = arr[high];
    arr[high] = temp;

    const pivot = arr[high];
    var i = low;

    for (low..high) |j| {
        if (arr[j] <= pivot) {
            const t = arr[i];
            arr[i] = arr[j];
            arr[j] = t;
            i += 1;
        }
    }

    const t = arr[i];
    arr[i] = arr[high];
    arr[high] = t;

    return i;
}

fn medianOfThree(comptime T: type, arr: []T, low: usize, high: usize) usize {
    const mid = low + (high - low) / 2;

    if (arr[low] > arr[mid]) {
        const t = arr[low];
        arr[low] = arr[mid];
        arr[mid] = t;
    }
    if (arr[low] > arr[high]) {
        const t = arr[low];
        arr[low] = arr[high];
        arr[high] = t;
    }
    if (arr[mid] > arr[high]) {
        const t = arr[mid];
        arr[mid] = arr[high];
        arr[high] = t;
    }

    return mid;
}

test "quick sort" {
    var arr = [_]i32{ 5, 2, 8, 1, 9, 3 };
    sort(i32, &arr);
    try std.testing.expectEqualSlices(i32, &[_]i32{ 1, 2, 3, 5, 8, 9 }, &arr);
}
